ci, packaging, meta: modernize Python packaging for Sopel (#2328)
sopel-irc/sopel: CONTRIBUTING.md

  the name of the thing you're changing in at the beginning of the message,
  followed by a colon: the plugin name for plugins, "docs" for documentation
  files, "coretasks" for `coretasks.py`, "db" for the database feature, etc.
* Python files should always have `from __future__ import annotations`
  as the first line after the module docstring.

Running from source
-------------------

Sopel is your typical Python project: you can install it from source, edit

> the code using standard tooling, and take advantage of your favourite code

Reviewer: I _think_ American spellings are pretty consistent throughout the docs. Not necessarily a preference, but I don't think anyone wants to go through all the documentation and change _everything_ to the UK spellings.

Author: Oh, it's "favorite", right? You'll have to point me out these things, because I'm not sure I'm even aware of all of them.

Author: ```suggestion
the code using standard tooling, and take advantage of your favorite code
```

editor.

Assuming you are using standard tools:

> * Create a virtualenv; you can use `virtualenv` directly, `virtualenvwrapper`,
>   or any tools that allows you to create and manage your virtualenv.

Reviewer: ```suggestion
* Create a virtualenv.
  * You can use `virtualenv` directly, `virtualenvwrapper`, or any tool that
    allows you to create and manage your virtualenv.
```

* Fork Sopel's repository, and clone your fork locally.
* Activate your virtualenv, and `cd` into your clone's folder.
* Ensure you have the latest version of `pip`, and install `wheel`.

> * Install Sopel from source as an editable install. If you don't use a project
>   manager (like Poetry or Hatch), you can do that with `pip install -e .` from
>   within the clone's directory, where there is a `pyproject.toml` file.

Reviewer: ```suggestion
* Install Sopel from source as an editable install.
  * If you don't use a project manager (like Poetry or Hatch), you can do
    that with `pip install -e .` from within the clone's directory, where
    there is a `pyproject.toml` file.
```

> * Install developments tools used to run tests and linters; there is a specific

Reviewer: ```suggestion
* Install development tools used to run tests and linters; there is a specific
```

Author: You can see me try. :upside_down_face:

  requirement file you can use with `pip install -U -r dev-requirements.txt`.
* Run `make qa` to run linters and tests.
* Run `make cleandoc` or just `make docs` to build the documentation locally.

> At this point, you are all set (at least for the Python environment), all

Reviewer: ```suggestion
At this point, you are all set, at least for the Python environment; all
```

> that's left to you is to configure your code editor the way you like, read and

Reviewer: ```suggestion
that's left is for you to configure your code editor the way you like, and read and
```

I'm intentionally not making this suggestion bigger because I'm on the fence about "and read and write". Need another brainwave.

write some code and documentation.

> Using branch
> ------------

Reviewer: ```suggestion
Using branches
--------------
```

As previously stated, you should fork Sopel's repository to implement your

> change and commits them. Moreover, we advise you to work from your own branch.

Reviewer: ```suggestion
changes and commit them. Moreover, we advise you to work from your own branch.
```

Additionally, could be useful to specify that each change/feature should be in its own branch.

To setup your local clone, we suggest the following steps:

```
$ cd path/to/your/dev/folder
$ mkdir sopel
$ cd sopel/
$ git clone git@github.com:<USERNAME>/sopel.git .
$ git remote add upstream git@github.com:sopel-irc/sopel.git
$ git checkout -b <your-branch-name>
```

With that workflow, you have your local clone with a link to the upstream
repository, and a branch to work on. Once you are done with your work, you can
commit, and push to your repository.

You'll probably need, at some point, to update your repository with the new
commits from upstream. We suggest the following steps:

```
$ git checkout master
$ git fetch upstream
$ git rebase upstream/master
$ git push
```

If you never pushed on your own `master` branch, you should not need to force
the push, which is why we recommend to work on your own branch.

> Said working can be updated with rebase:

Reviewer: What is meant by "Said working", so I can suggest a more grammatically correct alternative to this line?

Author: Said working... copy? Yeah, I think it's working copy. The goal is to talk about the local repository.

Author: ```suggestion
Said branch can be updated with rebase:
```

  the name of the thing you're changing in at the beginning of the message,
  followed by a colon: the plugin name for plugins, "docs" for documentation
  files, "coretasks" for `coretasks.py`, "db" for the database feature, etc.
* Python files should always have `from __future__ import annotations`
  as the first line after the module docstring.

Running from source
-------------------

Sopel is your typical Python project: you can install it from source, edit
the code using standard tooling, and take advantage of your favorite code
editor.

Assuming you are using standard tools:

* Create a virtualenv.

>   * You can use `virtualenv` directly, `virtualenvwrapper`, or any tool that

Reviewer 2: May want to also mention `venv` here for a stdlib option, and a link to the [PyPA](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/#creating-a-virtual-environment) guide might be helpful for contributors who aren't already familiar.

Reviewer: Hm, the PyPA guide says "virtualenv (for Python 2)" 🤔

Are `virtualenv` and `venv` close enough to feature parity that we can just recommend the stdlib option (`venv`) for newbies "or any other appropriate tool" if someone already has a preferred one?

Reviewer 2: The main difference is that `venv` provides an environment for whatever Python it's running in, and `virtualenv` has more ambition for managing multiple versions.

I would definitely be inclined towards recommending "`venv` or other tools" here, the users who want something other than the standard option probably don't need the docs to guide them to it anyway.

Author: I'd rather say:

> You can use `virtualenv` directly (or the `venv` built-in), `virtualenvwrapper` to manage multiple projects, or any tool that [...]

Running from source
-------------------

Sopel is your typical Python project: you can install it from source, edit
the code using standard tooling, and take advantage of your favorite code
editor.

Assuming you are using standard tools:

* Create a virtualenv.
  * You can use `virtualenv` directly, `virtualenvwrapper`, or any tool that
    allows you to create and manage your virtualenv.
* Fork Sopel's repository, and clone your fork locally.
* Activate your virtualenv, and `cd` into your clone's folder.
* Ensure you have the latest version of `pip`, and install `wheel`.
* Install Sopel from source as an editable install.
  * If you don't use a project manager (like Poetry or Hatch), you can do
    that with `pip install -e .` from within the clone's directory, where
    there is a `pyproject.toml` file.
* Install development tools used to run tests and linters; there is a specific
  requirement file you can use with `pip install -U -r dev-requirements.txt`.
* Run `make qa` to run linters and tests.
* Run `make cleandoc` or just `make docs` to build the documentation locally.

At this point, you are all set, at least for the Python environment; all

> that's left is for you to configure your code editor the way you like, and read and

Reviewer: Line _looks_ too long (but I didn't count columns, just compared to above).

write some code and documentation.

Using branches
--------------

As previously stated, you should fork Sopel's repository to implement your
changes and commit them. Moreover, we advise you to work from your own branch.

To setup your local clone, we suggest the following steps:

```

> $ cd path/to/your/dev/folder
> $ mkdir sopel
> $ cd sopel/
> $ git clone git@github.com:<USERNAME>/sopel.git .

Reviewer: For a subfolder named "sopel", this can be accomplished in fewer steps by:

```suggestion
$ cd path/to/your/dev/folder
$ git clone git@github.com:<USERNAME>/sopel.git
$ cd sopel/
```

$ git remote add upstream git@github.com:sopel-irc/sopel.git
$ git checkout -b <your-branch-name>
```

With that workflow, you have your local clone with a link to the upstream
repository, and a branch to work on. Once you are done with your work, you can
commit, and push to your repository.

You'll probably need, at some point, to update your repository with the new
commits from upstream. We suggest the following steps:

```
$ git checkout master
$ git fetch upstream

> $ git rebase upstream/master

Reviewer: Why not `git merge --ff-only upstream/master`? Unless I'm way off base about how rebase works, using it here may change SHAs on the downstream fork because `pick`ing a commit _copies_ it (and applies a new committer date). I've never seen someone suggest using `rebase` for this task before. 🤔

```suggestion
$ git merge --ff-only upstream/master
```

True, a straight merge risks adding a merge commit, but `--ff-only` aborts if the merge isn't a fast-forward.

Then again, if the user truly hasn't touched their own master branch, the nuclear option is fine too 😁:

```suggestion
$ git reset --hard upstream/master
```

Author: Ha ah. I do `git rebase upstream/master` all the time, because if you don't commit on your master branch, then what rebase does it take all the commit from upstream/master, and apply all the commits that exists only in your version. Therefore, only *your* commits are changed.

Basically, it's like `merge --ff-only`.

Reviewer: That'll count as "way off base about how rebase works", then. 😅

$ git push
```

If you never pushed on your own `master` branch, you should not need to force
the push, which is why we recommend to work on your own branch.

Said branch can be updated with rebase:

```
$ git checkout <your-branch-name>
$ git rebase master
```

To squash your commits, you can use interactive mode with
`git rebase -i master`.

> In both case, if you configured

Reviewer: Ach.

```suggestion
In both cases, if you configured
```

  the name of the thing you're changing in at the beginning of the message,
  followed by a colon: the plugin name for plugins, "docs" for documentation
  files, "coretasks" for `coretasks.py`, "db" for the database feature, etc.
* Python files should always have `from __future__ import annotations`
  as the first line after the module docstring.

Running from source
-------------------

Sopel is your typical Python project: you can install it from source, edit
the code using standard tooling, and take advantage of your favorite code
editor.

Assuming you are using standard tools:

* Create a virtualenv.
  * You can use `virtualenv` (or the `venv` built-in) directly,
    `virtualenvwrapper`, or any tool that allows you to create and manage your
    virtualenvs.

>   * Your project manager may create the virtualenv for you, be sure to check

Reviewer: Comma splice bad.

```suggestion
  * Your project manager may create the virtualenv for you; be sure to check
```

    the documentation of the tools you are using.
* Fork Sopel's repository, and clone your fork locally.
* Activate your virtualenv, and `cd` into your clone's folder.

> * Ensure you have the latest version of `pip`, and install `wheel`.

Reviewer: Is explicitly installing `wheel` still necessary? It and `setuptools` are declared as `build-system` dependencies in `pyproject.toml`.

Author: `wheel` is never necessary, it is only convenient: if installed, it means that doing `pip install -r dev-requirements.txt` for example will be able to generate wheels for the downloaded packages that don't provide them, even tho `pip install -e .` wasn't done yet.

So yeah, I consider that good advice anyway! It's like upgrading `pip` and installing `pytz`: these should be almost automatic when starting a new virtualenv.
